### PR TITLE
Don't send default cache header for 301 redirects

### PR DESCRIPTION
**Pull request by e-moe**
| Q | A |
| --- | --- |
| Branch? | "master" |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | yes |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #17139 |
| License | MIT |
| Doc PR | see comments from https://github.com/symfony/symfony/issues/17139 |

### CODE REVIEW

File: src/Symfony/Component/HttpFoundation/RedirectResponse.php [hunk 1]
```diff
@@ -41,6 +41,10 @@ public function __construct($url, $status = 302, $headers = array())
         if (!$this->isRedirect()) {
             throw new \InvalidArgumentException(sprintf('The HTTP status code is not a redirect ("%s" given).', $status));
         }
+
+        if (301 == $status && !array_key_exists('cache-control', $headers)) {
```
Comment (e-moe) [inline src/Symfony/Component/HttpFoundation/RedirectResponse.php]: I'm not sure if it's correct place for this code. Looking forward for review

Comment (fabpot) [inline src/Symfony/Component/HttpFoundation/RedirectResponse.php]: I would instead move this into the main Response class just before sending the headers to the client.

Comment (e-moe) [inline src/Symfony/Component/HttpFoundation/RedirectResponse.php]: @fabpot, in Response class headers already computed and contains unnecessary 'no-cache' directive. We can't check wether is was set automatically or explicitly by user. see `\Symfony\Component\HttpFoundation\ResponseHeaderBag::computeCacheControlValue`

**Comment by e-moe** [comment]
probably we can add `$withDefaultCacheControl` parameter to `\Symfony\Component\HttpFoundation\ResponseHeaderBag::__construct` and do something like this:

```
        if ($withDefaultCacheControl && !isset($this->headers['cache-control'])) {
            $this->set('Cache-Control', '');
        }
```

then, in `Response` constructor:

`$this->headers = new ResponseHeaderBag($headers, 301 != $status);` or, even better check against array like `$this->headers = new ResponseHeaderBag($headers, in_array($status, self::$useDefaultCacheControl));`

**Comment by fabpot** [comment]
I'm going to merge this as is.

**Comment by fabpot** [comment]
Thank you @e-moe.

**Pull request merged**